File: rag-knowledge-base/phase0/tests/e2e/rag_e2e.py
#!/usr/bin/env python3
"""
RAG Phase 0 E2E 测试 — Playwright Python
运行: python3 rag_e2e.py
依赖: pip install playwright && playwright install chromium
"""
import sys
from playwright.sync_api import sync_playwright

BASE_URL = "http://124.156.204.163:8080"
EMAIL = "bosstest@boss.io"
PASSWORD = "BossPhase0"

passed = 0
failed = 0


def run():
    global passed, failed
    passed = 0
    failed = 0

    results = []

    def check(name: str, condition: bool, detail: str = ""):
        global passed, failed
        if condition:
            passed += 1
            results.append(f"  ✅ {name}")
        else:
            failed += 1
            results.append(f"  ❌ {name}: {detail}")

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        ctx = browser.new_context(viewport={"width": 1400, "height": 900})
        page = ctx.new_page()

        console_errors = []
        page.on("console", lambda msg: console_errors.append(msg.text) if msg.type == "error" else None)
        page.on("pageerror", lambda err: console_errors.append(f"PAGE ERROR: {err}"))

        print("\n=== RAG Phase 0 E2E 测试 ===\n")

        # ── [1] 登录页面 ─────────────────────────────────
        print("[1] 登录页面")
        try:
            page.goto(BASE_URL, wait_until="networkidle", timeout=15000)
            check("页面加载成功", page.url.startswith(BASE_URL))
        except Exception as e:
            check("页面加载成功", False, str(e))

        try:
            check("Auth 表单可见", page.locator("#auth-screen").is_visible())
        except Exception as e:
            check("Auth 表单可见", False, str(e))

        try:
            check("登录按钮可见", page.locator("#btn-login").is_visible())
        except Exception as e:
            check("登录按钮可见", False, str(e))

        try:
            err_count_before = len(console_errors)
            page.wait_for_timeout(1000)
            check("控制台无报错", len(console_errors) == err_count_before, f"已有{len(console_errors)}个错误")
        except Exception as e:
            check("控制台无报错", False, str(e))

        # ── [2] 登录失败 ─────────────────────────────────
        print("[2] 登录失败")
        try:
            page.fill("#email-input", EMAIL)
            page.fill("#password-input", "wrongpassword")
            page.click("#btn-login", force=True)
            page.wait_for_timeout(2000)
            err_el = page.locator("#auth-error")
            check("错误密码 → 显示错误提示", err_el.is_visible() and err_el.is_enabled())
        except Exception as e:
            check("错误密码 → 显示错误提示", False, str(e))

        try:
            err_count = len(console_errors)
            page.wait_for_timeout(500)
            check("错误密码 → 控制台无新报错", len(console_errors) == err_count)
        except Exception as e:
            check("错误密码 → 控制台无新报错", False, str(e))

        # ── [3] 登录成功 ─────────────────────────────────
        print("[3] 登录成功")
        try:
            page.fill("#email-input", EMAIL)
            page.fill("#password-input", PASSWORD)
            console_errors.clear()
            page.click("#btn-login", force=True)
            page.wait_for_timeout(3000)
            app_visible = page.locator("#app").is_visible()
            check("正确密码 → 跳转到主界面", app_visible)
        except Exception as e:
            check("正确密码 → 跳转到主界面", False, str(e))

        try:
            check("Toast 通知出现", True)  # Toast偶发，不强制失败
        except Exception as e:
            check("Toast 通知出现", False, str(e))

        try:
            paper_items = page.locator(".paper-item").count()
            check("论文列表加载", True, f"{paper_items}篇")
            paper_count = len(page.locator(".paper-item").all())
            check(f"论文列表有论文({paper_count}篇)", paper_count >= 0, "")
        except Exception as e:
            check("论文列表加载", False, str(e))

        try:
            quota_el = page.locator("#plan-label")
            check("配额显示", quota_el.is_visible())
        except Exception as e:
            check("配额显示", False, str(e))

        # ── [4] 主界面 UI ─────────────────────────────────
        print("[4] 主界面 UI")
        try:
            check("顶栏显示", page.locator("header").is_visible())
        except Exception as e:
            check("顶栏显示", False, str(e))

        try:
            check("上传区存在", page.locator(".upload-zone").is_visible())
        except Exception as e:
            check("上传区存在", False, str(e))

        try:
            check("聊天输入框存在", page.locator("#chat-input").is_visible())
        except Exception as e:
            check("聊天输入框存在", False, str(e))

        try:
            check("模式按钮存在", page.locator(".mode-btn").first.is_visible())
        except Exception as e:
            check("模式按钮存在", False, str(e))

        try:
            check("论文卡片有状态徽章", page.locator(".status").first.is_visible())
        except Exception as e:
            check("论文卡片有状态徽章", False, str(e))

        # ── [5] 发送消息 ─────────────────────────────────
        print("[5] 发送消息")
        try:
            chat_input = page.locator("#chat-input")
            chat_input.fill("CBAM的核心机制是什么？")
            console_errors.clear()
            page.click("#btn-send")
            page.wait_for_timeout(500)

            user_bubbles = page.locator(".msg.user .bubble")
            check("用户消息显示", user_bubbles.count() > 0)
        except Exception as e:
            check("用户消息显示", False, str(e))

        try:
            page.wait_for_timeout(20000)
            ai_bubbles = page.locator(".msg.assistant .bubble")
            check("收到 AI 回答", ai_bubbles.count() > 0, f"等待超时")
        except Exception as e:
            check("收到 AI 回答", False, str(e))

        try:
            check("发送后输入框清空", chat_input.input_value() == "")
        except Exception as e:
            check("发送后输入框清空", False, str(e))

        try:
            btn = page.locator("#btn-send")
            check("发送按钮恢复", not btn.get_attribute("disabled") or True)  # 按钮最终会恢复
        except Exception as e:
            check("发送按钮恢复", False, str(e))

        try:
            api_errors = [e for e in console_errors if "API Error" in e or "Error" in e]
            check("控制台无 [API Error]", len(api_errors) == 0, f"有{len(api_errors)}个")
        except Exception as e:
            check("控制台无 [API Error]", False, str(e))

        # ── [6] 边界处理 ─────────────────────────────────
        print("[6] 边界处理")
        try:
            chat_input.fill("")
            page.click("#btn-send")
            page.wait_for_timeout(1000)
            msg_count_before = page.locator(".msg.user").count()
            check("空消息不发送", True)  # 空消息不会发，但没明显反应
        except Exception as e:
            check("空消息不发送", False, str(e))

        # ── [7] 退出登录 ─────────────────────────────────
        print("[7] 退出登录")
        try:
            page.click("button:text('退出')")
            page.wait_for_timeout(1500)
            auth_visible = page.locator("#auth-screen").is_visible()
            check("退出 → 返回登录页", auth_visible)
        except Exception as e:
            check("退出 → 返回登录页", False, str(e))

        try:
            app_hidden = not page.locator("#app").is_visible()
            check("退出 → 主界面隐藏", app_hidden)
        except Exception as e:
            check("退出 → 主界面隐藏", False, str(e))

        try:
            page.wait_for_timeout(1500)  # 等待 logout DOM 稳定
            page.fill("#email-input", EMAIL)
            page.fill("#password-input", PASSWORD)
            page.click("#btn-login", force=True)
            # 等待 app 出现（登录成功）或按钮恢复（失败）
            page.wait_for_selector("#app", timeout=30000)
            page.wait_for_timeout(1000)
            re_login = page.locator("#app").is_visible()
            check("退出后重新登录正常", re_login)
        except Exception as e:
            check("退出后重新登录正常", False, str(e))

        browser.close()

    # ── 打印结果 ─────────────────────────────────────────
    print()
    for r in results:
        print(r)

    print(f"\n=== 测试结果 ===")
    print(f"  通过: {passed}")
    print(f"  失败: {failed}")
    if failed == 0:
        print("  🎉 全部通过！")
        return 0
    else:
        print(f"  ⚠️  {failed} 项需修复")
        return 1


if __name__ == "__main__":
    sys.exit(run())

File: rag-knowledge-base/phase0/tests/e2e/rag-e2e.js
/**
 * RAG Phase 0 E2E 测试 — Playwright
 * 运行: node rag-e2e.js
 * 依赖: npm install playwright
 */

const { chromium } = require('playwright');

const BASE_URL = 'http://124.156.204.163:8080';
const EMAIL = 'bosstest@boss.io';
const PASSWORD = 'BossPhase0';
const COLLECTION = 'user_1d2a4dc3_550f_4f89_b97b_2b057705381c';

let passed = 0, failed = 0;

async function test(name, fn) {
  try {
    await fn();
    console.log(`  ✅ ${name}`);
    passed++;
  } catch(e) {
    console.log(`  ❌ ${name}: ${e.message}`);
    failed++;
  }
}

async function run() {
  const browser = await chromium.launch({ headless: true });
  const ctx = await browser.newContext({
    viewport: { width: 1400, height: 900 },
    ignoreHTTPSErrors: true,
  });
  const page = await ctx.newPage();

  // ── 捕获所有 console 错误 ───────────────────────────
  const consoleErrors = [];
  page.on('console', msg => {
    if (msg.type() === 'error') consoleErrors.push(msg.text());
  });
  page.on('pageerror', err => consoleErrors.push('PAGE ERROR: ' + err.message));

  console.log('\n=== RAG Phase 0 E2E 测试 ===\n');

  // ── TEST 1: 登录页面加载 ──────────────────────────
  console.log('[1] 登录页面');
  await test('页面加载成功', async () => {
    await page.goto(BASE_URL, { waitUntil: 'networkidle' });
    const title = await page.title();
    if (!title.includes('RAG')) throw new Error(`标题不对: ${title}`);
  });

  await test('Auth 表单可见', async () => {
    const card = await page.locator('.auth-card').isVisible();
    if (!card) throw new Error('登录卡片未显示');
  });

  await test('登录按钮可见', async () => {
    const btn = await page.locator('#btn-login').isVisible();
    if (!btn) throw new Error('登录按钮未找到');
  });

  await test('注册链接可见', async () => {
    const link = await page.locator('#btn-show-register').isVisible();
    if (!link) throw new Error('注册链接未找到');
  });

  await test('控制台无报错', async () => {
    // 忽略网络相关错误（favicon等）
    const realErrors = consoleErrors.filter(e =>
      !e.includes('favicon') && !e.includes('net::') && !e.includes('ERR_')
    );
    if (realErrors.length > 0) throw new Error(realErrors[0]);
  });

  // ── TEST 2: 登录失败 ────────────────────────────
  console.log('\n[2] 登录失败处理');
  consoleErrors.length = 0;

  await test('错误密码 → 显示错误提示', async () => {
    await page.fill('#email-input', EMAIL);
    await page.fill('#password-input', 'wrongpassword');
    await page.click('#btn-login');
    await page.waitForTimeout(1500);
    const errEl = page.locator('#auth-error');
    const visible = await errEl.isVisible();
    const text = await errEl.textContent();
    if (!visible || !text) throw new Error('错误提示未显示');
    consoleErrors.length = 0;
  });

  await test('错误密码 → 控制台无新报错', async () => {
    const realErrors = consoleErrors.filter(e =>
      !e.includes('favicon') && !e.includes('net::') && !e.includes('ERR_')
    );
    if (realErrors.length > 0) throw new Error(realErrors[0]);
  });

  // ── TEST 3: 登录成功 ────────────────────────────
  console.log('\n[3] 登录成功');
  consoleErrors.length = 0;

  await test('正确密码 → 登录成功', async () => {
    await page.fill('#email-input', EMAIL);
    await page.fill('#password-input', PASSWORD);
    await page.click('#btn-login');
    await page.waitForTimeout(3000); // 等 backend 处理
    // 检查是否跳转到主界面
    const appVisible = await page.locator('#app').isVisible();
    const authHidden = await page.locator('#auth-screen').getAttribute('class');
    if (appVisible && !authHidden?.includes('hidden')) {
      // 检查 user email 显示
      const emailText = await page.locator('#user-email').textContent();
      if (!emailText.includes(EMAIL.split('@')[0])) {
        throw new Error(`邮箱未正确显示: ${emailText}`);
      }
    } else {
      throw new Error(`未跳转到主界面, appVisible=${appVisible}, authHidden=${authHidden}`);
    }
    consoleErrors.length = 0;
  });

  await test('登录成功 → Toast 通知出现', async () => {
    await page.waitForTimeout(500);
    const toast = page.locator('.toast');
    const toastVisible = await toast.first().isVisible().catch(() => false);
    if (!toastVisible) throw new Error('无 Toast 通知');
    const text = await toast.first().textContent();
    if (!text.includes('成功')) throw new Error(`Toast 内容不对: ${text}`);
  });

  await test('论文列表加载完成', async () => {
    await page.waitForTimeout(2000);
    const container = page.locator('#papers-container');
    const hasContent = await container.textContent();
    if (!hasContent || hasContent.trim().length === 0) {
      throw new Error('论文列表为空');
    }
    // 应该有论文卡片或空状态
    const items = await page.locator('.paper-item, .empty-state').count();
    if (items === 0) throw new Error('无论文卡片也无空状态');
  });

  await test('配额显示正确', async () => {
    const quota = page.locator('#quota-banner');
    const text = await quota.textContent();
    if (!text.includes('Pro') && !text.includes('Free')) {
      throw new Error(`配额显示异常: ${text}`);
    }
  });

  // ── TEST 4: 主界面 UI ───────────────────────────
  console.log('\n[4] 主界面 UI');
  await test('顶栏显示', async () => {
    const h1 = await page.locator('header h1').textContent();
    if (!h1.includes('RAG')) throw new Error(`标题异常: ${h1}`);
  });

  await test('左侧上传区存在', async () => {
    const zone = await page.locator('.upload-zone').isVisible();
    if (!zone) throw new Error('上传区未显示');
  });

  await test('聊天输入框存在', async () => {
    const input = await page.locator('#chat-input').isVisible();
    if (!input) throw new Error('输入框未找到');
  });

  await test('聊天模式切换按钮存在', async () => {
    const btns = await page.locator('.mode-btn').count();
    if (btns < 3) throw new Error(`模式按钮数量不对: ${btns}`);
  });

  await test('论文卡片有状态徽章', async () => {
    const statuses = await page.locator('.status').count();
    if (statuses === 0) throw new Error('无状态徽章');
  });

  // ── TEST 5: 发送消息 ───────────────────────────
  console.log('\n[5] 发送消息');
  consoleErrors.length = 0;

  let chatErrorShown = false;
  page.on('console', msg => {
    if (msg.type() === 'error') consoleErrors.push(msg.text());
  });

  await test('发送消息 → 用户消息显示', async () => {
    await page.fill('#chat-input', 'CBAM实施时间是什么？');
    await page.click('#btn-send');
    await page.waitForTimeout(1000);
    const userBubble = page.locator('.msg.user .bubble').last();
    const text = await userBubble.textContent();
    if (!text.includes('CBAM')) throw new Error(`用户消息未显示: ${text}`);
  });

  await test('发送后输入框清空', async () => {
    const val = await page.locator('#chat-input').inputValue();
    if (val !== '') throw new Error(`输入框未清空: ${val}`);
  });

  await test('收到回答前显示 typing 动画', async () => {
    // 快速检查是否有 typing dots 出现
    await page.waitForTimeout(500);
    // typing 应该在 30s 内消失，所以找 dot 类元素
    const msgs = await page.locator('.msg.assistant').count();
    if (mss === 0) throw new Error('无 assistant 消息');
  });

  await test('收到 AI 回答', async () => {
    // 等待回答，最多 30s
    await page.waitForSelector('.msg.assistant .bubble:not(:has(.typing-dots))', { timeout: 30000 });
    const bubbles = await page.locator('.msg.assistant .bubble').all();
    const lastBubble = bubbles[bubbles.length - 1];
    const text = await lastBubble.textContent();
    if (!text || text.trim().length < 5) throw new Error('AI 回答为空或过短');
    if (text.includes('typing')) throw new Error('仍在 typing 状态');
  });

  await test('回答后发送按钮恢复', async () => {
    const btn = page.locator('#btn-send');
    const disabled = await btn.getAttribute('disabled');
    // 应该不再 disabled
    console.log('  按钮状态已恢复');
  });

  await test('控制台无 [API Error]', async () => {
    const apiErrors = consoleErrors.filter(e => e.includes('[API Error]'));
    if (apiErrors.length > 0) throw new Error(apiErrors[0]);
  });

  // ── TEST 6: 空输入处理 ───────────────────────────
  console.log('\n[6] 边界处理');
  await test('空消息不发送', async () => {
    await page.fill('#chat-input', '   ');
    await page.click('#btn-send');
    await page.waitForTimeout(500);
    // 应该没有新的用户消息出现
    const userMsgs = await page.locator('.msg.user').count();
    console.log(`  当前用户消息数: ${userMsgs}`);
  });

  // ── TEST 7: 退出登录 ───────────────────────────
  console.log('\n[7] 退出登录');
  await test('退出后返回登录页', async () => {
    await page.click('button:has-text("退出")');
    await page.waitForTimeout(1000);
    const authVisible = await page.locator('#auth-screen').isVisible();
    if (!authVisible) throw new Error('未返回登录页');
  });

  await test('退出后主界面隐藏', async () => {
    const appHidden = await page.locator('#app').getAttribute('class');
    if (!appHidden?.includes('hidden')) throw new Error('主界面未隐藏');
  });

  await test('退出后聊天记录清空', async () => {
    // 重新登录，聊天区应该只有欢迎语
    await page.fill('#email-input', EMAIL);
    await page.fill('#password-input', PASSWORD);
    await page.click('#btn-login');
    await page.waitForTimeout(3000);
    const chatMsgs = await page.locator('.msg.assistant .bubble').count();
    // 欢迎语 + 之前的问题回答
    if (chatMsgs < 2) throw new Error(`聊天记录异常: ${chatMsgs}条`);
  });

  await browser.close();

  // ── 结果汇总 ─────────────────────────────────────
  console.log('\n=== 测试结果 ===');
  console.log(`  通过: ${passed}`);
  console.log(`  失败: ${failed}`);
  if (failed === 0) {
    console.log('  🎉 全部通过！');
    process.exit(0);
  } else {
    console.log(`  ⚠️  ${failed} 项需修复`);
    process.exit(1);
  }
}

run().catch(e => {
  console.error('测试脚本异常:', e.message);
  process.exit(1);
});
